_posts/2015-11-09-Generators.md

```diff
@@ -0,0 +1,250 @@
+
+```javascript
+function *test() {
+  yield 1;
+  return "complete";
+}
+const gen = test();
+console.log(gen.next()); // { done: false, value: 1 }
+console.log(gen.next()); // { done: true, value: "complete" }
+console.log(gen.next()); // { done: true, value: undefined }
+```
+
+Because of this behavior, manually iterating poses typing difficulties. Let's try to take the first 10 values from the `nats` generator through manual iteration:
+
+```javascript
+const gen = nats();
+const take10: number[] = [];
+for (let i = 0; i < 10; i++) {
+  const { done, value } = gen.next();
+  if (done) {
+    break;
+  } else {
+    take10.push(value); // error!
+  }
+}
+```
+
+```
+test.js:13
+ 13:   const { done, value } = gen.next();
+                               ^^^^^^^^^^ call of method `next`
+ 17:     take10.push(value); // error!
+                     ^^^^^ undefined. This type is incompatible with
+ 11: const take10: number[] = [];
+                   ^^^^^^ number
+```
+
+Flow is complaining that `value` might be `undefined`. This is because the type of `value` is `Yield | Return | void`, which simplifies in the instance of `nats` to `number | void`. We can introduce a dynamic type test to convince Flow of the invariant that `value` will always be `number` when `done` is false.
+
+```javascript
+const gen = nats();
+const take10: number[] = [];
+for (let i = 0; i < 10; i++) {
+  const { done, value } = gen.next();
+  if (done) {
+    break;
+  } else {
+    if (typeof value === "undefined") {
+      throw new Error("`value` must be a number.");
+    }
+    take10.push(value); // no error
+  }
+}
+```
+
+There is an [open issue](https://github.com/facebook/flow/issues/577) which would make the dynamic type test above unnecessary, by using the `done` value as a sentinel to refine a tagged union. That is, when `done` is `true`, Flow would know that `value` is always of type `Yield` and otherwise of type `Return | void`.
+
+Even without the dynamic type test, this code is quite verbose and it's hard to see the intent. Because generators are also iterable, we can also use `for...of` loops:
+
+```javascript
+const take10: number[] = [];
+let i = 0;
+for (let nat of nats()) {
+  if (i === 10) break;
+  take10.push(nat);
+  i++;
+}
+```
+
+That's much better. The `for...of` looping construct ignores completion values, so Flow understands that `nat` will always be `number`. Let's generalize this pattern further using generator functions:
+
+```javascript
+function *take<T>(n: number, xs: Iterable<T>): Iterable<T> {
+  if (n <= 0) return;
+  let i = 0;
+  for (let x of xs) {
+    yield x;
+    if (++i === n) return;
+  }
+}
+
+for (let n of take(10, nats())) {
+  console.log(n); // 0, 1, 2, 3, 4, 5, 6, 7, 8, 9
+}
+```
+
+Note that we explicitly annotated the parameters and return type of the `take` generator. This is necessary to ensure Flow understands the fully generic type.
```

Comment:
Yeah, definitely. I had this originally and cut it for length, but I think
it's a useful thing to cover. Easy to re-add.

On Wed, Nov 11, 2015 at 9:33 AM, Jeff Morrison <notifications@github.com>
wrote:

> In _posts/2015-11-09-Generators.md
> <https://github.com/facebook/flow/pull/1063#discussion_r44575210>:
>
> > +```javascript
> > +function *take<T>(n: number, xs: Iterable<T>): Iterable<T> {
> > +  if (n <= 0) return;
> > +  let i = 0;
> > +  for (let x of xs) {
> > +    yield x;
> > +    if (++i === n) return;
> > +  }
> > +}
> > +
> > +for (let n of take(10, nats())) {
> > +  console.log(n); // 0, 1, 2, 3, 4, 5, 6, 7, 8, 9
> > +}
> > +```
> > +
> > +Note that we explicitly annotated the parameters and return type of the `take` generator. This is necessary to ensure Flow understands the fully generic type.
>
> Could you expand a bit on what Flow does if you hadn't been so explicit
> with annotations?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/facebook/flow/pull/1063/files#r44575210>.
>